### PR TITLE
Warn when a user tries to set a profile name that is unlikely to be valid

### PR DESCRIPTION
**Pull request by bhanu011**
Need an early review before merging

### CODE REVIEW

File: cmd/minikube/cmd/config/util_test.go [hunk 1]
```diff
@@ -111,3 +112,26 @@ func TestIsAddonAlreadySet(t *testing.T) {
...
+		},
+	}
+
+	for _, test := range testCases {
+		profileNam := test.profileName
+		expectedMsg := test.expectErr
+
+		err := ValidateProfile(profileNam)
+		fmt.Println(err.Error())
+		if err.Error() != expectedMsg {
```
Comment (tstromberg) [inline cmd/minikube/cmd/config/util_test.go]: While it's OK to assert that an error satisfies a property such as including the name of the profile, error strings should not be compared directly: it's too fragile. See https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully for more details. 

There are a couple of ways to do this, but here's my recommendation:

- Give this error a type in util.go. For instance:

```go
type ErrProfileDoesNotExist struct {
  Name string
}

func (e ErrProfileDoesNotExist) Error() string {
	return fmt.Sprintf("profile does not exist: %q", e.Name)
}
```

- Update `ValidateProfile` to return `ErrProfileDoesNotExist{Name: profile}`
- Update this test to check the error against this type:

```
switch err := errors.Cause(err).(type) {
  case ErrProfileDoesNotExist:
    return nil
  case default:
    t.Errorf("ValidateProfile error: %v, want: %v")
```

This way you get rich error messages and the ability to compare errors. Hope this helps!

**Comment by tstromberg** [comment]
@minikube-bot OK to test

File: cmd/minikube/cmd/config/util.go [hunk 2]
```diff
@@ -226,3 +227,31 @@ func EnableOrDisableStorageClasses(name, val string) error {
...
+	var profileFound bool
+	validProfiles, invalidProfiles, err := pkgConfig.ListProfiles()
+	if err != nil {
+		exit.WithError("Not able to retrieve profile list", err)
+	}
+
+	// handling invalid profiles
+	for _, invalidProf := range invalidProfiles {
+		if profile == invalidProf.Name {
+			return errors.New("You are trying to switch to invalid profile")
```
[the remark this answers is not in the document]
Comment (bhanu011) [reply cmd/minikube/cmd/config/util.go]: I modify according to suggestion

File: cmd/minikube/cmd/config/util.go [hunk 3]
```diff
@@ -226,3 +227,31 @@ func EnableOrDisableStorageClasses(name, val string) error {
...
+
+	// valid profiles if found, setting profileFound to true
+	for _, prof := range validProfiles {
+		if prof.Name == profile {
+			profileFound = true
+			break
+		}
+	}
+	if !profileFound {
+		return errors.New("Profile you are trying to switch is not found")
```
[the remark this answers is not in the document]
Comment (bhanu011) [reply cmd/minikube/cmd/config/util.go]: I modify according to suggestion

File: cmd/minikube/cmd/config/util.go [hunk 4]
```diff
@@ -226,3 +227,31 @@ func EnableOrDisableStorageClasses(name, val string) error {
 
 	return EnableOrDisableAddon(name, val)
 }
+
+//ValidateProfile checks if the profile user is trying to switch exists, else throws error
```
[the remark this answers is not in the document]
Comment (bhanu011) [reply cmd/minikube/cmd/config/util.go]: I modify according to suggestion

File: cmd/minikube/cmd/config/util.go [hunk 5]
```diff
@@ -226,3 +227,31 @@ func EnableOrDisableStorageClasses(name, val string) error {
...
+func ValidateProfile(profile string) error {
+	var profileFound bool
```
[the remark this answers is not in the document]
Comment (bhanu011) [reply cmd/minikube/cmd/config/util.go]: I thought profileFound := false will be redundant since `var profileFound bool` declaration is doing it.

**Comment by bhanu011** [comment]
> First, this looks good. I'd love to see a test for a passing TestValidateProfile.
> 
> I hope you don't mind, but I've added a bunch of tips for writing idiomatic Go & Kubernetes code. I know that this PR is a WIP, but I hope you are able to find the comments useful. Thank you for proposing this fix!

Thanks a lot for review comments mentioning why it has to be in that way. I am learning a lot from this contribution

File: cmd/minikube/cmd/config/profile.go [hunk 6]
```diff
@@ -47,6 +47,12 @@ var ProfileCmd = &cobra.Command{
 		profile := args[0]
 		if profile == "default" {
 			profile = "minikube"
+		} else {
+			// not validating when it is default profile
+			errProfile, ok := ValidateProfile(profile)
+			if !ok && errProfile != nil {
+				out.FailureT(errProfile.Msg)
+			}
```
Comment (bhanu011) [inline cmd/minikube/cmd/config/profile.go]: @tstromberg is this validation here fine?

**Comment by bhanu011** [comment]
Fixes #4883

**Comment by RA489** [comment]
@minikube-bot OK to test

**Comment by tstromberg** [comment]
Integration tests appear to be unrelated flakes:

*HyperKit_macOS*: `TestContainerd/GvisorRestart:  StartHost: Temporary Error: Error configuring auth on host: OS type not recognized`
*VirtualBox_Linux*: `TestPersistence: certificate is valid for 192.168.99.102, 10.96.0.1, 10.0.0.1, not 192.168.99.103`
*VirtualBox_macOS*: `TestProxy/ProxyDashboard: z_proxy_test.go:145: failed to read url: timeout after 3m0s`

File: cmd/minikube/cmd/config/profile.go [hunk 7]
```diff
@@ -47,6 +47,12 @@ var ProfileCmd = &cobra.Command{
 		profile := args[0]
 		if profile == "default" {
 			profile = "minikube"
+		} else {
+			// not validating when it is default profile
+			errProfile, ok := ValidateProfile(profile)
+			if !ok && errProfile != nil {
+				out.FailureT(errProfile.Msg)
```
Comment (medyagh) [inline cmd/minikube/cmd/config/profile.go]: this PR looks good ! but we need to add a message to the end user, how to create a profile !

this PR will make it minikube start to be the only way to create a profile.

**Comment by medyagh** [comment]
@bhanu011 thank for taking this PR ! I just compiled your code and I tried to do minikube ssh on a not existing profile (px):

```
 $ ./out/minikube ssh -p px
⚠️  minikube 1.3.1 is available! Download it: https://github.com/kubernetes/minikube/releases/tag/v/1.3.1
💡  To disable this notice, run: 'minikube config set WantUpdateNotification false'

💣  Error getting host: Machine does not exist for api.Exists(px)

😿  Sorry that minikube crashed. If this was unexpected, we would love to hear from you:
👉  https://github.com/kubernetes/minikube/issues/new/choose
```

it seems like the your ValidarteProfile func was not called !
I supsect you need to add the validate profile other than only the profile cmd

**Comment by medyagh** [comment]
one good place to do test this would be integration tests.

I suggest adding an integration tests to https://github.com/kubernetes/minikube/blob/master/test/integration/fn_profile.go


 in which you can try to run "minikube logs" on  a profile that doesn't exist and expect the output of the message to be yours.

and also please make sure to error to the user how to create a profile.
you could say, you can create a profile using "minikube start -p pname"

for example 

```
mk := NewMinikubeRunner(t, "not_existing_profile")
stdout,stderr, err: = mk.RunCommand("ssh")
// parse the stdout and stderr and expect to see your vaildate profile error message
```

here is more info on integration testing :
https://minikube.sigs.k8s.io/docs/contributing/testing/

one tip on running intregation test:

you can run only One Test by this command:

``` env TEST_ARGS="-minikube-start-args=--vm-driver=hyperkit -test.run TestFunctional" make integration
```

**Comment by tstromberg** [comment]
@bhanu011 - Would you like any help to move this PR along?

**Comment by medyagh** [comment]
@bhanu011 are you still interested in finishing up this PR?

**Comment by woodcockjosh** [comment]
@bhanu011 `minikube profile options` will set the profile to `options` instead of listing the options for setting the profile. Will this PR resolve that?

**Comment by sharifelgamal** [comment]
@minikube-bot OK to test

**Pull request merged**